// @selfxyz/qrcode/dist/chunk-OHEJ4ULG.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _chunkJJXEZUAKcjs = require('./chunk-JJXEZUAK.cjs');

// components/LED.tsx
var _react = require('react'); var _react2 = _interopRequireDefault(_react);
var green = "#31F040";
var blue = "#424AD8";
var gray = "#95a5a6";
var LED = ({
  size = 8,
  connectionStatus = _chunkJJXEZUAKcjs.QRcodeSteps.DISCONNECTED
}) => {
  const getColor = () => {
    if (connectionStatus >= _chunkJJXEZUAKcjs.QRcodeSteps.MOBILE_CONNECTED) {
      return green;
    } else if (connectionStatus >= _chunkJJXEZUAKcjs.QRcodeSteps.WAITING_FOR_MOBILE) {
      return blue;
    } else {
      return gray;
    }
  };
  return /* @__PURE__ */ _react2.default.createElement(
    "div",
    {
      style: {
        width: `${size}px`,
        height: `${size}px`,
        borderRadius: "50%",
        backgroundColor: getColor(),
        boxShadow: `0 0 ${size * 1.5}px ${getColor()}`,
        transition: "all 0.3s ease",
        marginBottom: "8px"
      }
    }
  );
};
var LED_default = LED;



exports.LED_default = LED_default;
