// @selfxyz/common/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSolidityPackedUserContextData = exports.calculateUserIdentifierHash = exports.buildSMT = exports.genMockIdDocAndInitDataParsing = exports.generateMockDSC = exports.genMockIdDoc = exports.parseDscCertificateData = exports.parseCertificateSimple = exports.getLeafDscTree = exports.getLeafCscaTree = exports.findStartPubKeyIndex = exports.generateNullifier = exports.generateCommitment = exports.Hash = exports.brutforceSignatureAlgorithmDsc = exports.REDIRECT_URL = exports.generateCircuitInputsVCandDisclose = exports.generateCircuitInputsRegister = exports.generateCircuitInputsDSC = exports.genAndInitMockPassportData = exports.initPassportDataParsing = exports.getCircuitNameFromPassportData = exports.formatMrz = exports.getSKIPEM = exports.initElliptic = void 0;
const constants_js_1 = require("./src/constants/constants.js");
Object.defineProperty(exports, "REDIRECT_URL", { enumerable: true, get: function () { return constants_js_1.REDIRECT_URL; } });
const parseCertificateSimple_js_1 = require("./src/utils/certificate_parsing/parseCertificateSimple.js");
Object.defineProperty(exports, "parseCertificateSimple", { enumerable: true, get: function () { return parseCertificateSimple_js_1.parseCertificateSimple; } });
const passport_js_1 = require("./src/utils/passports/passport.js");
Object.defineProperty(exports, "findStartPubKeyIndex", { enumerable: true, get: function () { return passport_js_1.findStartPubKeyIndex; } });
Object.defineProperty(exports, "generateCommitment", { enumerable: true, get: function () { return passport_js_1.generateCommitment; } });
Object.defineProperty(exports, "generateNullifier", { enumerable: true, get: function () { return passport_js_1.generateNullifier; } });
const parseDscCertificateData_js_1 = require("./src/utils/passports/passport_parsing/parseDscCertificateData.js");
Object.defineProperty(exports, "parseDscCertificateData", { enumerable: true, get: function () { return parseDscCertificateData_js_1.parseDscCertificateData; } });
const trees_js_1 = require("./src/utils/trees.js");
Object.defineProperty(exports, "getLeafCscaTree", { enumerable: true, get: function () { return trees_js_1.getLeafCscaTree; } });
Object.defineProperty(exports, "getLeafDscTree", { enumerable: true, get: function () { return trees_js_1.getLeafDscTree; } });
const genMockIdDoc_js_1 = require("./src/utils/passports/genMockIdDoc.js");
Object.defineProperty(exports, "genMockIdDoc", { enumerable: true, get: function () { return genMockIdDoc_js_1.genMockIdDoc; } });
Object.defineProperty(exports, "generateMockDSC", { enumerable: true, get: function () { return genMockIdDoc_js_1.generateMockDSC; } });
Object.defineProperty(exports, "genMockIdDocAndInitDataParsing", { enumerable: true, get: function () { return genMockIdDoc_js_1.genMockIdDocAndInitDataParsing; } });
const brutForceDscSignature_js_1 = require("./src/utils/passports/passport_parsing/brutForceDscSignature.js");
Object.defineProperty(exports, "brutforceSignatureAlgorithmDsc", { enumerable: true, get: function () { return brutForceDscSignature_js_1.brutforceSignatureAlgorithmDsc; } });
const trees_js_2 = require("./src/utils/trees.js");
Object.defineProperty(exports, "buildSMT", { enumerable: true, get: function () { return trees_js_2.buildSMT; } });
var elliptic_js_1 = require("./src/utils/certificate_parsing/elliptic.js");
Object.defineProperty(exports, "initElliptic", { enumerable: true, get: function () { return elliptic_js_1.initElliptic; } });
var csca_js_1 = require("./src/utils/csca.js");
Object.defineProperty(exports, "getSKIPEM", { enumerable: true, get: function () { return csca_js_1.getSKIPEM; } });
var format_js_1 = require("./src/utils/passports/format.js");
Object.defineProperty(exports, "formatMrz", { enumerable: true, get: function () { return format_js_1.formatMrz; } });
var circuitsName_js_1 = require("./src/utils/circuits/circuitsName.js");
Object.defineProperty(exports, "getCircuitNameFromPassportData", { enumerable: true, get: function () { return circuitsName_js_1.getCircuitNameFromPassportData; } });
const Hash = __importStar(require("./src/utils/hash.js"));
exports.Hash = Hash;
const hash_js_1 = require("./src/utils/hash.js");
Object.defineProperty(exports, "calculateUserIdentifierHash", { enumerable: true, get: function () { return hash_js_1.calculateUserIdentifierHash; } });
Object.defineProperty(exports, "getSolidityPackedUserContextData", { enumerable: true, get: function () { return hash_js_1.getSolidityPackedUserContextData; } });
__exportStar(require("./src/constants/countries.js"), exports);
__exportStar(require("./src/constants/constants.js"), exports);
__exportStar(require("./src/utils/appType.js"), exports);
__exportStar(require("./src/utils/scope.js"), exports);
var passport_js_2 = require("./src/utils/passports/passport.js");
Object.defineProperty(exports, "initPassportDataParsing", { enumerable: true, get: function () { return passport_js_2.initPassportDataParsing; } });
var genMockPassportData_js_1 = require("./src/utils/passports/genMockPassportData.js");
Object.defineProperty(exports, "genAndInitMockPassportData", { enumerable: true, get: function () { return genMockPassportData_js_1.genAndInitMockPassportData; } });
var generateInputs_js_1 = require("./src/utils/circuits/generateInputs.js");
Object.defineProperty(exports, "generateCircuitInputsDSC", { enumerable: true, get: function () { return generateInputs_js_1.generateCircuitInputsDSC; } });
Object.defineProperty(exports, "generateCircuitInputsRegister", { enumerable: true, get: function () { return generateInputs_js_1.generateCircuitInputsRegister; } });
Object.defineProperty(exports, "generateCircuitInputsVCandDisclose", { enumerable: true, get: function () { return generateInputs_js_1.generateCircuitInputsVCandDisclose; } });
